rfcs/0052-dynamic-ids.md

```diff
@@ -0,0 +1,112 @@
+---
+feature: dynamic-ids
+start-date: 2019-09-05
+author: Silvan Mosberger
+co-authors: (find a buddy later to help our with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: https://github.com/NixOS/nixpkgs/pull/65698
+---
+
+# Summary
+[summary]: #summary
+
+A lot of NixOS modules are [assigning static uids/gids](https://github.com/NixOS/nixpkgs/blob/044cc701c23ede96355eb1dc997985b4dfac0372/nixos/modules/misc/ids.nix#L36) to their users. This has resulted in less than 90 static ids left in the reserved range from 0 to 400.
+
+This RFC deprecates the practice of doing that and instead suggests to
+- If applicable use systemd's `DynamicUser`
+- Otherwise let NixOS assign dynamic persistent ids, which happens automatically when `users.users.<name?>.uid`/`users.groups.<name?>.gid` is not set. For users, `users.users.<name?>.isSystemUser` should be set so that only uids under 1000 are used by NixOS services.
+
+Only in special circumstances are static ids allowed anymore.
+
+# Motivation
+[motivation]: #motivation
+
+We are running out of static ids in the reserved range from 0 to 400. If services continue to reserve ids for themselves this will run out eventually.
+
+In addition, a central list of ids is annoying to maintain and leads to merge conflicts.
+
+# Detailed design
+[design]: #detailed-design
+
+## Documentation updates
+
+New best practices for declaring users are to be documented. This includes sections like the following
+
+### Using `DynamicUser` for declaring users
+
+If the service is fit for `DynamicUser`, this is the preferred solution, as it doesn't require any persistent ids. However this only works well for services whose state is self-contained. As soon as the services data is needed somewhere else, this approach can fall flat. `DynamicUser` also enables a bunch of options restricting the service (see `man systemd.exec` for details), some of which might need to be disabled in order for it to work.
```

Comment:
I still have some hopes that at some point we will have a shiftfs in linux mainline: https://github.com/linuxkit/linuxkit/tree/master/projects/shiftfs
Then we could have `DynamicUser` for all services because it removes the overhead of adapting filesystem uids at service startup.

Reply:
https://discuss.linuxcontainers.org/t/lxd-3-12-has-been-released/4483

> This is a feature we’ve been looking forward to for years and that we are really excited to finally see coming to completion. shiftfs allows for an unprivileged container experience that doesn’t need any shifting of the filesystem, instead having the kernel do it on the fly.
> 
> This requires kernel support through the shiftfs filesystem which is currently a custom patchset that will be present in the Ubuntu 19.04 kernel.
> 
> LXD automatically detects support for this and will transparently start using it whenever possible.

Sounds like maybe Ubuntu adopted a `shiftfs` kernel patch.